src/routes.tsx

```diff
@@ -0,0 +1,346 @@
+import {
+  matchPath, Outlet,
+} from 'react-router-dom';
+import { PageWrap } from '@edx/frontend-platform/react';
+
+import RouteErrorBoundary from './components/app/routes/RouteErrorBoundary';
+import Root from './components/app/Root';
+import Layout from './components/app/Layout';
+import { makeRootLoader } from './components/app/routes/loaders';
+import NotFoundPage from './components/NotFoundPage';
+
+/**
+ * Returns the route loader function if a queryClient is available; otherwise, returns null.
+ */
+export function getRouteLoader(makeRouteLoaderFn: Types.MakeRouteLoaderFunction, queryClient?: Types.QueryClient) {
+  if (!queryClient) {
+    return undefined;
+  }
+  return makeRouteLoaderFn(queryClient);
+}
+
+/**
+ * Returns the routes nested under the enterprise slug prefix.
+ */
+function getEnterpriseSlugRoutes(queryClient?: Types.QueryClient) {
```

Comment:
[question/curious] If the loaders are now gating for the query client should the query client be an optional argument? Are there any instances where we don't expect the query client to exist before the routes are loaded and that being an acceptable outcome (as to not gate our loaders with the following code snippet)? 
``` jsx
if (!authenticatedUser || !queryClient) {
      return null;
    }
```

Should we just make the queryClient a non optional as an additional gating step for `getEnterpriseSlugRoutes`? (This might just be a gap in my TS knowledge too)

Reply:
`getRoutes` is called in two places: `createAppRouter` (internal) and `env.config.js` (external). Because `env.config` is calling `getRoutes` _without_ passing a `queryClient` (i.e., one does not exist yet), the `getRoutes` and child functions like `getEnterpriseSlugRoutes` need to conditionally handle when `queryClient` is missing. The interaction with `getRouteLoader` checks whether `queryClient` is present and, if so, passed it to the route loader.

The shared route loader types themselves (i.e.,`Types.MakeRouteLoaderFunction`) defines any route loader creator functions as conditionally accepting a `queryClient` as not all route loaders in the app use a query client.

  To be able to remove the `if (!authenticatedUser || !queryClient) {` gating in a route loader, we could create a new loader-specific type that extends the base `Types.MakeRouteLoaderFunction` type, telling TypeScript that this particular loader requires `queryClient`.

```tsx
// base makeRouteLoader type
export type MakeRouteLoaderFunction = (queryClient?: QueryClient) => RouteLoaderFunction;

// extended loader-specific makeRouteLoaderType that requires queryClient
type MakeRootLoaderFunction = (queryClient: Types.QueryClient) => Types.RouteLoaderFunction;

const makeRootLoader: MakeRootLoaderFunction = function makeRootLoader(queryClient) {
  // no more check for `!queryClient` needed
}
```

Reply:
Thank you for this explanation! Leave unresolved for knowledge sharing purposes IMO.